File: source/python/1. Two Sum.py
# import itertools


class Solution:
    # # Most Naive Solution
    # def twoSum(self, nums: list[int], target: int) -> list[int]:
    #     all_comb = itertools.combinations(nums, 2)

    #     for pairs in all_comb:
    #         result = pairs[0] + pairs[1]

    #         if result == target:
    #             if pairs[0] != pairs[1]:
    #                 return [nums.index(pairs[0]), nums.index(pairs[1])]

    #             first = nums.index(pairs[0])
    #             second = nums.index(pairs[1], first + 1)
    #             return [first, second]

    #     return []

    # Efficient Solution using Hash Table
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        hash_table = {}

        for index, num in enumerate(nums):
            if target - num in hash_table:
                return [hash_table[target - num], index]

            hash_table[num] = index

        return []

File: source/typescript/1. Two Sum.ts
function twoSum(nums: number[], target: number): number[] {
    const hashmap = new Map<number, number>();

    for (let i = 0; i < nums.length; i++) {
        if (hashmap.has(target - nums[i]))
            return [hashmap.get(target - nums[i])!, i];

        hashmap.set(nums[i], i);
    }

    return [];
};
